### ISSUE TITLE
Arrayed keyword arguments seem to have regressed

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#6418]
## Environment

* SuperCollider version: current dev (d52b052)
* Operating system: n/a
* Other details (Qt version, audio driver, etc.): n/a

## Steps to reproduce

```supercollider
f = { |a, b ... c|
	[a, b, c].postln;
};

f.(c: 5)
```

If I roll back to commit 976098f, the result is `[nil, nil, 5]` (though this is incorrect in itself, as `c` should be wrapped in an array: `[nil, nil, [5]]` would be expected).

Current dev head (d52b052):

```
WARNING: keyword arg 'c' not found in call to function.
[nil, nil, []]
```

## Expected vs. actual behavior

I discovered this when attempting to open a MixingBoard (ddwMixerChannel quark), using a formulation that I've been using for years:

```
MixingBoard(mixers: mixer, mixer, mixer, mixer...)
```

... and I got an empty window, because `mixers` never got the values.

BTW: The keyword-argument warning occurs for Function:value, but *not* for `*new` calls -- the MixingBoard case fails silently (creates an empty window, with no warning about why).

So... we cannot guarantee that there is no user code that expects to be able to keyword-address "..."-style args. Perhaps this is not found in the main class library, but this is one instance of breakage in an unexpected place, in an extension. There could be others.

IMO the ideal solution would be for `f` to get `[5]` for `c`.

If that's not possible, then we should let users know that there's potential for old code to break here.

**Comment by jamshark70** [comment]
> Consider this example, what should f return?
> 
> ```
> f = { |...a| a }
> f.(1, 2, a: 10)
> ```

I'd approach this question by going back to first principles, rolling back to the older version and testing an even simpler case:

```
f = { |a| a };
f.(1, a: 2);
```

... where `a` becomes 2. So the precedent, from SC prior to the new keyword argument features, is that the keyword argument overrides previously-specified positional values. (Also, the last-specified keyword argument overrides previously-specified keyword values: in `f.(1, a: 2, a: 3)`, a == 3.)

^^ This, btw, is documented: https://doc.sccode.org/Reference/Messages.html#Keyword%20Arguments -- "keyword value of 1200 overrides the positional arg value of 800" in `SinOsc.ar(800, freq: 1200)`. The fact that it's documented makes it more of a concrete principle IMO.

Extending that principle to `...a`, then, `a` should be `[10]` because keyword values override positional ones: after `a:`, the 1 and 2 don't exist anymore.

I guess I'm also not worried about that case, because it's inherently a bad idea for anyone to write extra args that they know will be overridden by keywords. But, per documentation, there is a defined behavior for it.

> Because we need to pass the keyword through performList, wrapping (your example) c in an array and producing [nil, nil, [5]] is not possible.

I was going to say something else, but on second thought, I think this may be a documentation issue. Documentation needs to emphasize the importance of disambiguation. The performList cases that you mention appear to be extremely difficult (if not impossible) to disambiguate -- therefore users should be warned that distributing arrayed values to multiple arguments *plus* keyword arguments is something to avoid.

```
Object.performList { |...args | _ObjectPerformList }
```

It might help here to rename `args` to something that a user would never accidentally collide with, say, `|... listOfArgumentValues|`. (Nobody's gonna type all that routinely...) `|... args|` is fairly common practice in user code; it might be worth it for the main library to bend over a bit further backward to avoid conflicts.

Tested:

```
f = { |a, b, args| [a, b, args] }

f.(*[1, 2], args: 10) // [1, 2, 10] OK! (in new version)
```

So there's a good reason to change these core methods away from `args`, and I can't think of a good reason not to.

**Comment by JordanHendersonMusic** [comment]
> ^^ This, btw, is documented: https://doc.sccode.org/Reference/Messages.html#Keyword%20Arguments -- "keyword value of 1200 overrides the positional arg value of 800" in SinOsc.ar(800, freq: 1200). The fact that it's documented makes it more of a concrete principle IMO.

!!! I did not know that, nor did I know there was some documentation on kw args. Right now it does the opposite and posts this warning:
`Ignoring duplicate keyword 'a' at position 0 found in function call`

I think the warning should stay, but I'll change the behaviour to do what the docs say.

I'll make another issue for this.

---

> therefore users should be warned that distributing arrayed values to multiple arguments plus keyword arguments is something to avoid.

I do not understand this.

The way I see it, and the way the code worked before and now, is that every thing after `...` is *not* a normal argument, but a placeholder for anything that isn't caught by the preceding arguments.

Simply changing the name and restoring the old behaviour will cause many other issues, notably in object prototyping, alongwith in all places where `doesNotUnderstand` or `performList` is called. That PR fixed all of these issues with keyword arguments.

Based on the months of work that both I and @telephon did exploring this area, I think your original usage was abusing a bug and should be changed. Exactly how this is communicated to quark authors and users I do not no, but I suspect that the number of people who relied on this bug is minimal. 


---

Would you mind just clarifying what the *intention* of the follow call is, and that you've copied it correctly, as you can't have normal arguments **after** a keyword arguments as per the Bison file.

```supercollider
*new { arg name = "Mixer", skin ... mixers;
   ^super.new.init(name, skin, mixers);
}

MixingBoard(mixers: mixer1, mixer2, mixer3, mixer4...)
```

```
| expr '.' '(' arglist1 optkeyarglist ')' blocklist
```
The optional keyword argument list always happens after normal args.

**Comment by JordanHendersonMusic** [comment]
Just to quote @telephon who might have more thoughts on this.

>"It's a bug, not a feature"

https://github.com/supercollider/supercollider/pull/6297#issuecomment-2096803171

**Comment by jamshark70** [comment]
> > therefore users should be warned that distributing arrayed values to multiple arguments plus keyword arguments is something to avoid.
> 
> Edit: To clarify, I think it is a bug that you could pass and override as a keyword argument something intended to capture any uncaught arguments.

Let me just suggest that this is perhaps not disagreement at all, but rather agreement: It's a bad idea to mix styles here. Programming practices that result in ambiguities that can be resolved only by arbitrary fiat are by definition bad programming practices.

**Comment by jamshark70** [comment]
> Would you mind just clarifying what the intention of the following call is, and that you've copied it correctly, as you can't have normal arguments after a keyword arguments as per the Bison file.

Yeah... I think I mis-stated. I seldom use MixingBoard directly anymore. It's indeed a syntax error to list non-keyword arguments after keyword arguments, so I must have been mistaken about that.

In the past, I would have done `MixingBoard(mixers: oneMixer)` or `MixingBoard(mixers: [aMixer, anotherMixer])`. MixingBoard flattens `mixers`, so it wouldn't matter to this class if `mixers` were `[aMixer, anotherMixer]` or `[[aMixer, anotherMixer]]`. Now, neither of those formulations works.

**Comment by JordanHendersonMusic** [comment]
Dealing with this inside the method seems like the best (`asArray` or something).

Do you think this issue can be closed, or perhaps renamed focusing on documentation?

**Comment by telephon** [comment]
To me it seems that this is a good occasion to remove some boundary cases that are confusing both in use and implementation. Unfortunately, this means a bit of breakage. I may be biased, but it seems like worth it, right?

**Comment by JordanHendersonMusic** [comment]
I'm closing this issue because it seems to be resolved. Being able to assign to the identifier used to capture any unnamed arguments throws away information and creating a work around to achieve the old behaviour is possible.
